include/eos/fitting/ceres_nonlinear.hpp

```diff
@@ -420,31 +445,437 @@ std::array<T, 3> get_shape_point(const morphablemodel::PcaModel& shape_model,
+        /* Templates: CostFunctor, num residuals, blendshape-coeffs */
+        auto* blendshapes_prior_cost =
+                new ceres::AutoDiffCostFunction<fitting::PriorCost,blendshapes_num, blendshapes_num>(
+                        new fitting::PriorCost(blendshapes_num, weigth));
+        problem->AddResidualBlock(blendshapes_prior_cost, nullptr, &blendshape_coefficients[0]);
+
+        for (int i = 0; i < blendshapes_num; ++i) {
+            problem->SetParameterLowerBound(&blendshape_coefficients[0], i, 0.0);
+        }
+    }
+
+
+    #if EOS_CERES_USE_OPENCV == true
+    /**
+     * Add residual block with image cost function to the passed problem
+     *
+     * @param[in, out] camera camera to fit
+     * @param[in] image image to fit
+     */
+    template<bool use_perspective>
+    void add_image_cost_function(CameraParameters<use_perspective>& camera, const cv::Mat& image) {
+        // Add a residual for each vertex:
+        for (int i = 0; i < morphable_model->get_shape_model().get_data_dimension() / 3; ++i) {
+            // Templates: CostFunctor, Residuals: [R, G, B], camera rotation (quaternion),
+            //            camera translation & focal length, shape-coeffs, bs-coeffs, colour coeffs
+
+            auto* cost_function = new ceres::AutoDiffCostFunction<fitting::ImageCost, 3, 4,
+                    get_num_cam_params(use_perspective),
+                    shapes_num, blendshapes_num, color_coeffs_num>(
+                    new fitting::ImageCost(*morphable_model, *blendshapes, image, i, use_perspective));
+
+            problem->AddResidualBlock(cost_function, nullptr, &camera.rotation_quaternion[0],
+                                     &camera.translation_and_intrinsics[0], &shape_coefficients[0],
+                                     &blendshape_coefficients[0], &colour_coefficients[0]);
+        }
+    }
+
+    /**
+     * Add residual block with shape prior cost function to the overall problem
+     *
+     * @param[in] weight constant to multiply all colours
+     * @param[in, out] upper and lower bound of each colour_coefficient
+     */
+    void add_image_prior_cost_function(double weigth = 35.0,
+                                       double color_coeff_limit = 3.0) {
+        // Templates: CostFunctor, num residuals, colour-coeffs
+        auto* colour_prior_cost =
+                new ceres::AutoDiffCostFunction<fitting::PriorCost, color_coeffs_num, color_coeffs_num>(
+                        new fitting::PriorCost(color_coeffs_num, weigth));
+
+        problem->AddResidualBlock(colour_prior_cost, nullptr, &colour_coefficients[0]);
+        for (int i = 0; i < color_coeffs_num; ++i) {
+            problem->SetParameterLowerBound(&colour_coefficients[0], i, -color_coeff_limit);
+            problem->SetParameterUpperBound(&colour_coefficients[0], i, color_coeff_limit);
+        }
+    }
+    #endif
+
+    /**
+     * Block shapes fitting
+     */
+    void block_shapes_fitting() {
```

Comment:
With "block" here you mean "set constant"? I think "block" could be confused with the term "parameter block" from Ceres. Perhaps name it `set_shape_coefficients_constant()`?
(Same applies to the similar functions below)

Reply:
OK